app/supplier_utils.py

```diff
@@ -37,3 +44,80 @@ def check_supplier_role(role, supplier_id):
         abort(400, "'supplierId' is required for users with 'supplier' role")
     elif role != 'supplier' and supplier_id:
         abort(400, "'supplierId' is only valid for users with 'supplier' role, not '{}'".format(role))
+
+
+def company_details_confirmed_if_required_for_framework(framework_slug, supplier_framework):
+    return (
+        supplier_framework.supplier.company_details_confirmed
+        if framework_slug == 'g-cloud-10' else
+        supplier_framework.application_company_details_confirmed
+    )
+
+
+def update_open_declarations_with_company_details(db, supplier_id, updater_json, specific_framework_slug=None):
+    """Expected to be called within a view"""
+    open_supplier_frameworks_query = SupplierFramework.query.filter(
+        SupplierFramework.supplier_id == supplier_id,
+    )
+
+    if specific_framework_slug:
+        open_supplier_frameworks_query = open_supplier_frameworks_query.filter(
+            SupplierFramework.framework.has(Framework.slug == specific_framework_slug),
+        )
+    else:
+        open_supplier_frameworks_query = open_supplier_frameworks_query.filter(
+            SupplierFramework.framework.has(Framework.status == 'open'),
+        )
+
+    # We need to ensure that SupplierFramework and Framework haven't changed during this transaction, so grab
+    # the affected rows with a for update lock.
+    open_supplier_frameworks = open_supplier_frameworks_query.options(lazyload('*')).with_for_update().all()
```

Comment:
Is the Framework going to be loaded here with `with_for_update`? I'm guessing it is because the query's above include a check on the Framework (namely its status) and so they will be loaded with the `lazyload` option? Otherwise would you need to explicitly exclude Framework from the lazyload option?

Reply:
This is the emitted statement, which selects from both tables, so they should both get locked.

```SQL
SELECT supplier_frameworks.supplier_id AS supplier_frameworks_supplier_id, supplier_frameworks.framework_id AS supplier_frameworks_framework_id, supplier_frameworks.declaration AS supplier_frameworks_declaration, supplier_frameworks.on_framework AS supplier_frameworks_on_framework, supplier_frameworks.agreed_variations AS supplier_frameworks_agreed_variations, supplier_frameworks.prefill_declaration_from_framework_id AS supplier_frameworks_prefill_declaration_from_framework_id, supplier_frameworks.application_company_details_confirmed AS supplier_frameworks_application_company_details_confirmed
FROM supplier_frameworks
WHERE supplier_frameworks.supplier_id = %(supplier_id_1)s AND (EXISTS (SELECT 1
FROM frameworks
WHERE frameworks.id = supplier_frameworks.framework_id AND frameworks.status = %(status_1)s)) AND (EXISTS (SELECT 1
FROM frameworks
WHERE frameworks.id = supplier_frameworks.framework_id AND frameworks.slug = %(slug_1)s)) FOR UPDATE
```